db-auto-backup.py

```diff
@@ -160,70 +160,154 @@ def get_backup_provider(container_names: Iterable[str]) -> Optional[BackupProvid
-                method="write",
-                desc=container.name,
-                disable=not SHOW_PROGRESS,
-            ) as f:
-                for stdout, _ in output:
-                    if stdout is None:
-                        continue
-                    f.write(stdout)
+def create_backup_file_name(container: Container, backup_provider: BackupProvider) -> Path:
+    """
+    Create a backup file name with a timestamp prefix and the container name.
+    """
+    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
+    container_name = container.name
+    return BACKUP_DIR / f"{timestamp}_{container_name}.{backup_provider.file_extension}{get_compressed_file_extension(COMPRESSION)}"
 
-        os.replace(backup_temp_file_path, backup_file)
 
-        if not SHOW_PROGRESS:
-            print(container.name)
+@pycron.cron(SCHEDULE)
+def backup(now: datetime) -> None:
+    print("Starting backup...")
 
-        backed_up_containers.append(container.name)
+    docker_client = docker.from_env()
 
-    duration = (datetime.now() - now).total_seconds()
-    print(
-        f"Backup of {len(backed_up_containers)} containers complete in {duration:.2f} seconds."
-    )
+    if is_swarm_mode():
+        print("Running in Swarm mode, adjusting container lookup...")
+        tasks = get_local_node_tasks()
+        backed_up_services = []
+
+        for task in tasks:
+            task_container_id = task['Status']['ContainerStatus']['ContainerID']
+            try:
+                container = docker_client.containers.get(task_container_id)
+            except docker.errors.NotFound:
+                continue
+
+            container_names = get_container_names(container)
+            backup_provider = get_backup_provider(container_names)
+            
+            if backup_provider is None:
+                continue
+
+            backup_file = create_backup_file_name(container, backup_provider)
+            backup_temp_file_path = BACKUP_DIR / temp_backup_file_name()
+
+            backup_command = backup_provider.backup_method(container)
+            _, output = container.exec_run(backup_command, stream=True, demux=True)
+
+            with open_file_compressed(
+                backup_temp_file_path, COMPRESSION
+            ) as backup_temp_file:
+                with tqdm.wrapattr(
+                    backup_temp_file,
+                    method="write",
+                    desc=task["ServiceID"],
+                    disable=not SHOW_PROGRESS,
+                ) as f:
+                    for stdout, _ in output:
+                        if stdout is None:
+                            continue
+                        f.write(stdout)
+
+            os.replace(backup_temp_file_path, backup_file)
+            backed_up_services.append(container.name)
+
+        duration = (datetime.now() - now).total_seconds()
+        print(f"Backup of {len(backed_up_services)} services complete in {duration:.2f} seconds.")
+    else:
```

Comment:
Question: What's the reason for the below duplication? Could the extra logic be handled in the existing single loop?